torch/csrc/autograd/profiler.cpp

```diff
@@ -8,24 +8,57 @@
 #include <string>
 #include <vector>
 
+#include <ATen/ThreadLocalState.h>
+
 namespace torch { namespace autograd { namespace profiler {
 
 namespace {
 
 CUDAStubs default_stubs;
 constexpr CUDAStubs* default_stubs_addr = &default_stubs;
-// constant initialization, so it is guaranteed to be initialized before
+// Constant initialization, so it is guaranteed to be initialized before
 // static initialization calls which may invoke registerCUDAMethods
 static CUDAStubs* cuda_stubs = default_stubs_addr;
 
-ProfilerState state = ProfilerState::Disabled;
+thread_local ProfilerConfig config(ProfilerState::Disabled, false);
 // Protects access all_event_lists_map.
 std::mutex all_event_lists_map_mutex;
 std::unordered_map<uint16_t, std::shared_ptr<RangeEventList>>
     all_event_lists_map;
 thread_local std::shared_ptr<RangeEventList> event_list;
 thread_local uint16_t thread_id;
 
+bool unused_ = []() {
+  at::ThreadLocalState::registerThreadLocalSetting(
+    at::ThreadLocalSetting::PROFILER,
+    []() {
+      return at::SettingValue{.pair =
+          {(int32_t)config.state, (int32_t)config.report_input_shapes}};
+    },
+    [](at::SettingValue v) {
+      // ThreadLocalState propagates profiler state across threads and
+      // automatically enables/disables profiling in async tasks.
+      // Note, that in case of a subtask profiling, we don't
+      // want to consolidate profiling results but let the original profiler
+      // keep running
+      if (v.pair.first != (int32_t)ProfilerState::Disabled) {
```

Comment:
can we ever end up in a situation that profiler was already enabled?

Reply:
(commented below)